morph.go

```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
+	if err != nil {
+		return nil, err
+	}
+
+	if mode == models.Down {
+		sortedMigrations := reverseSortMigrations(appliedMigrations)
+		downMigrations, err := findDownScripts(sortedMigrations, m.source.Migrations())
+		if err != nil {
+			return nil, err
+		}
+
+		diff := make([]*models.Migration, 0, len(downMigrations))
+		for i := 0; i < len(sortedMigrations); i++ {
+			diff = append(diff, downMigrations[sortedMigrations[i].Name])
+		}
+
+		return diff, nil
+	}
+
+	pendingMigrations, err := computePendingMigrations(appliedMigrations, m.source.Migrations())
+	if err != nil {
+		return nil, err
+	}
+
+	var diff []*models.Migration
+	for _, migration := range sortMigrations(pendingMigrations) {
+		if migration.Direction != models.Up {
+			continue
+		}
+		diff = append(diff, migration)
+	}
+
+	return diff, nil
+}
+
+func (m *Morph) GetOppositeMigrations(migrations []*models.Migration) ([]*models.Migration, error) {
+	var direction models.Direction
+	migrationsMap := make(map[string]*models.Migration)
+	for _, migration := range migrations {
+		if direction == "" {
+			direction = migration.Direction
+		}
+		// check if the migrations has the same direction
+		if direction != migration.Direction {
+			return nil, errors.New("migrations have different directions")
+		}
+
+		migrationsMap[migration.Name] = migration
+	}
+
+	rollbackMigrations := make([]*models.Migration, 0, len(migrations))
+	availableMigrations := m.source.Migrations()
+	for _, migration := range availableMigrations {
+		// skip if we have the same direction for the migration
+		// we are looking for opposite direction
+		if migration.Direction == direction {
+			continue
+		}
+
+		// we don't have the migration in the map
+		// so we can't rollback it
+		_, ok := migrationsMap[migration.Name]
+		if !ok {
+			continue
+		}
+
+		rollbackMigrations = append(rollbackMigrations, migration)
+	}
+
+	if len(migrations) != len(rollbackMigrations) {
+		return nil, errors.New("not all migrations have opposite migrations")
+	}
+
+	return rollbackMigrations, nil
+}
+
+// GeneratePlan returns the plan to apply these migrations and also includes
+// the safe rollback steps for the given migrations.
+func (m *Morph) GeneratePlan(migrations []*models.Migration) (*models.Plan, error) {
+	rollbackMigrations, err := m.GetOppositeMigrations(migrations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get opposite migrations: %w", err)
+	}
+
+	plan := models.NewPlan(migrations, rollbackMigrations)
+
+	return plan, nil
+}
+
+func (m *Morph) ApplyPlan(plan *models.Plan) error {
+	if err := plan.Validate(); err != nil {
+		return fmt.Errorf("invalid plan: %w", err)
+	}
+
+	revertMigrations := make([]*models.Migration, 0, len(plan.RevertMigrations))
+
+	for i := range plan.Migrations {
+		// add to the revert queue
+		for _, migration := range plan.RevertMigrations {
+			if migration.Name == plan.Migrations[i].Name {
+				revertMigrations = append(revertMigrations, migration)
+				break
+			}
+		}
```

Comment:
Any reason to get this separately again? In `NewPlan`, can't we validate that len(migrations) == len(reverseMIgrations), and the name of each migration at index `i` is the same? Then we don't need all this

Reply:
So the logic here is to have a stack of revert migrations. We push migrations here and then read in reverse. We are just making sure that we are setting the right order of what we have applied or attempted to apply.

I think we can lean on the order in the plan, but this is just safer to me.

Reply:
My mistake, see https://github.com/mattermost/morph/pull/17#discussion_r1022447007